### DISCUSSION TITLE
Request Enhancement: convex hull

### DISCUSSION THREAD

**Discussion by fmw42** [ImageMagick/ImageMagick#4862] [Development]
I would like to be able to list the 4 extreme points on the convex hull.

-define convex-hull:extremepoints="top(x,y) right(x,y) bottom(x,y) left(x,y)

or

```
-define convex-hull:top=(x,y)
-define convex-hull:right=(x,y)
-define convex-hull:bottom=(x,y)
-define convex-hull:left=(x,y)

```

where

```
top(x,y) is the x where y is minimum (use left-most x if more than one x has the same minimum y)
right(x,y) is the y where x is maximum (use top-most y if more than one y has the same maximum x)
bottom(x,y) is the x where y is maximum (use right-most x if more than one x has the same maximum y)
left(x,y) is the y where x is minimum (use bottom-most y if more than one y has the same minimum x)

```

**Comment by urban-warrior** [comment]
Not sure what you're asking.  Given the convex hull algorithm, where are these extreme points introduced into the algorithm?

**Reply by fmw42** [reply]
The resulting points for the convex hull will have a point that has min x, max x, min y and max y somewhere along the convex hull points. Find those extrement values (perhaps from the bounding box). Then find the corresponding points (x,y)  that has those extrement values. (i.e. the points of the convex hull that intersect with the bounding box of the convex hull)

**Comment by urban-warrior** [comment]
We need  an example.  Post an example image, the command you use to generate the convex hull, the points returned for the convex hull, a set of defines you want to use to modify the current convex hull results (-define convex-hull:top=(x,y)), and finally, how the convex hull image or points list changes given your defines.

**Comment by fmw42** [comment]
New string format:

`"%[convex-hull:extremes]" or "%[convex-hull:extreme-points]"`

would print

`"top(x,y) right(x,y) bottom(x,y) left(x,y)"`

where
```
top(x,y) is the x where y is minimum (use left-most x if more than one x has the same minimum y)
right(x,y) is the y where x is maximum (use top-most y if more than one y has the same maximum x)
bottom(x,y) is the x where y is maximum (use right-most x if more than one x has the same maximum y)
left(x,y) is the y where x is minimum (use bottom-most y if more than one y has the same minimum x)

```
Example:

`magick rose: -rotate 20 +repage -negate -threshold 10% rose_rot20_thresh.png`
![rose_rot20_thresh](https://user-images.githubusercontent.com/13458218/155900205-93f0feb9-e72f-4a27-9363-e7b42f586f9b.png)

whose convex hull is the following:
![rose_rot20_thresh_hull](https://user-images.githubusercontent.com/13458218/155900298-20c4e19e-df7e-4d51-b6f6-bf875bb99e28.png)



```
# list the convex hull and store points in variable
pts=`magick -quiet rose_rot20_thresh.png -format "%[convex-hull]\n" info: | tr " " "\n"`
echo "$pts"
17,1
23,3
78,23
82,25
82,26
81,29
68,65
67,67
66,68
60,66
5,46
1,44
1,43
2,40
7,26
15,4
16,2
17,1

```

Now find the following by sorting:

```
#top is x,y where y is min and if y min duplicates, then where x is min
echo "$pts" | sort -t "," -n -k 2,2 -k 1,1
17,1 <--- top
17,1
16,2
23,3
15,4
78,23
82,25
7,26
82,26
81,29
2,40
1,43
1,44
5,46
68,65
60,66
67,67
66,68

#right is x,y where x is max and if x max duplicates where y is min
echo "$pts" | sort -t "," -nr -k 1,1 -k 2,2
82,26
82,25 <--right
81,29
78,23
68,65
67,67
66,68
60,66
23,3
17,1
17,1
16,2
15,4
7,26
5,46
2,40
1,44
1,43

#bottom is x,y where y is max and if y max duplicates, then where x is max
echo "$pts" | sort -t "," -nr -k 2,2 -k 1,1
66,68 <--bottom
67,67
60,66
68,65
5,46
1,44
1,43
2,40
81,29
82,26
7,26
82,25
78,23
15,4
23,3
16,2
17,1
17,1

#left is x,y where x is min and if x max duplicates where y is max
echo "$pts" | sort -t "," -n -k 1,1 -k 2,2
1,43
1,44 <--left
2,40
5,46
7,26
15,4
16,2
17,1
17,1
23,3
60,66
66,68
67,67
68,65
78,23
81,29
82,25
82,26

```


So for the example above, the string format would compute and list the following for the convex hull extreme points

`17,1 82,25 66,68 1,44`

**Reply by urban-warrior** [reply]
Beta is building now, use `%[convex-hull:extreme-points]`.

**Comment by fmw42** [comment]
Seems to be working fine.  Thanks

magick rose: -rotate 20 +repage -negate -threshold 10% rose_rot20_thresh.png

magick rose_rot20_thresh.png -format "%[convex-hull:extreme-points]" info:
17,1 82,25 66,68 1,44
